pkg/provisioner/ironic/ironic.go

```diff
@@ -1175,17 +1195,43 @@ func (p *ironicProvisioner) Deprovision() (result provisioner.Result, err error)
 		result.RequeueAfter = deprovisionRequeueDelay
 		return result, nil
 
-	case nodes.Manageable, nodes.Enroll, nodes.Verifying:
+	case nodes.Enroll, nodes.Verifying, nodes.Adopting:
+		p.log.Info("re-registering to deprovision")
+		result.Dirty = true
+		result.RequeueAfter = deprovisionRequeueDelay
+		return result, nil
+
+	case nodes.Manageable:
+		// We can end up "manageable" under two conditions.  If we end
+		// up manageable because of a successful deprovisioning
+		// operation, when the node moves from cleaning to manageable,
+		// the instance_info structure is cleared out so we know we
+		// are done. If we end up here because we re-register a host,
+		// moving from verifying to manageable, then we have an image
+		// source and we need to adopt the node so we can try to
+		// deprovision all over again.
```

Comment:
Why exactly? Because you want to run cleaning? Otherwise you can just delete a node.

Reply:
If we think we should be deprovisioning, and we end up here with the image_source set, then we did not finish deprovisioning. So, yes, we still need to clean the host.